### PR TITLE
Single selection combo implementation

### LINKED ISSUE #9832: IgxCombo with single selection
## Question  
I have a potentially large list and I need a dropdown component with the ability to filter/search.
The IgxAutocomplete doesn't have the ability to scroll through all available values.
The IgxCombo would be perfect but it forces multi selection.
This behavior can be changed with `onSelectionChange` like described in the documentation but the value of the component is still an array.
Are there any plans to add a singleSelection mode where the value is not an array but a single object?
I think a single selection dropdown with filter is a quite common use case.

 * igniteui-angular version: 12.0.7
 * browser: Chrome

### PR DESCRIPTION
**Pull request by jackofdiamond5**
Closes #9832

### Additional information (check all that apply):
 - [ ] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [ ] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 1]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+     */
+    private getValueDisplayPairs(ids: any[]) {
+        return this.data.filter(entry => ids.indexOf(entry[this.valueKey]) > -1).map(e => ({
+            [this.valueKey]: e[this.valueKey],
+            [this.displayKey]: e[this.displayKey]
+        }));
+    }
+
+    /** Returns a string that should be populated in the combo's text box */
+    private concatDisplayText(selection: any[]): string {
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:1271]: do we need this in common? IMO simple combo does not need such method.

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 2]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+
+    /**
+     * Get current selection state
+     *
+     * @returns Array of selected items
+     * ```typescript
+     * let selectedItems = this.combo.selectedItems();
+     * ```
+     */
+    public selectedItems() {
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:1012]: this should become `selected`

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 3]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+    public set type(val: IgxInputGroupType) {
+        this._type = val;
+    }
+    /**
+     * An @Input property that controls whether the combo's search box
+     * should be focused after the `opened` event is called
+     * When `false`, the combo's list item container will be focused instead
+     */
+    @Input()
+    public autoFocusSearch = true;
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:451]: this should not go to common. simple combo has no search box in the drop down.

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 4]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+    }
+
+    /**
+     * An @Input property that enabled/disables filtering in the list. The default is `true`.
+     * ```html
+     * <igx-combo [filterable]="false">
+     * ```
+     */
+    @Input()
+    public filterable = true;
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:363]: do we need this for simple combo? Setting this to `false` there makes UX really terrible.

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 5]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+     * ```html
+     * <!--set-->
+     * <igx-combo [displayKey]='myDisplayKey'></igx-combo>
+     * ```
+     */
+    public get displayKey() {
+        return this._displayKey ? this._displayKey : this.valueKey;
+    }
+
+    /**
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:331]: dev docs should be placed on the getter

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 6]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+    @Input()
+    public get data(): any[] | null {
+        return this._data;
+    }
+    public set data(val: any[] | null) {
+        this._data = (val) ? val : [];
+    }
+
+    /**
+     * Combo value data source property.
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:290]: this is incomprehensible. Please change with something like this
_"Determines which column in the data source is used to determine the value."_

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 7]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+    @Input()
+    public valueKey: string = null;
+
+    @Input()
+    public set displayKey(val: string) {
+        this._displayKey = val;
+    }
+
+    /**
+     * Combo text data source property.
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:311]: same as `valueKey`

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 8]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+     * ```
+     *
+     * ```html
+     * <!--set-->
+     * <igx-combo [showSearchCaseIcon]='true'></igx-combo>
+     * ```
+     */
+
+    @Input()
+    public showSearchCaseIcon = false;
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:271]: `showSearchCaseIcon` does nothing in simple combo. We should either remove it from common or make it work.

File: projects/igniteui-angular/src/lib/combo/combo.common.ts [hunk 9]
```diff
@@ -27,6 +53,1244 @@ export interface IgxComboBase {
...
+     * let myComboSearchPlaceholder = this.combo.searchPlaceholder;
+     * ```
+     *
+     * ```html
+     * <!--set-->
+     * <igx-combo [searchPlaceholder]='newPlaceHolder'></igx-combo>
+     * ```
+     */
+    @Input()
+    public searchPlaceholder = 'Enter a Search Term';
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/combo/combo.common.ts:254]: this makes no sense to go in common. Search placeholder is used in the drop down input of the `igx-combo`. Simple combo has no such input. Now setting `searchPlaceholder` for simple combo works as `placeholder`. However setting both `searchPlaceholder` and `placeholder` will end up showing the `placeholder`.

File: projects/igniteui-angular/src/lib/simple-combo/simple-combo.component.html [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    </ng-container>
+    <ng-container ngProjectAs="igx-prefix">
+        <ng-content select="igx-prefix"></ng-content>
+    </ng-container>
+    <ng-container ngProjectAs="igx-hint, [igxHint]">
+        <ng-content select="igx-hint, [igxHint]"></ng-content>
+    </ng-container>
+
+    <input #comboInput igxInput [value]="value" (input)="handleInputChange($event)" (keyup)="handleKeyUp($event)"
+        (keydown)="handleKeyDown($event)" (blur)="onBlur()" [attr.placeholder]="placeholder || searchPlaceholder"
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/simple-combo.component.html:13]: either remove `placeholder` or `searchPlaceholder`. Having both these is contradictory.

File: projects/igniteui-angular/src/lib/simple-combo/simple-combo.component.ts [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+        }
+    }
+
+    /** @hidden @internal */
+    public getEditElement(): HTMLElement {
+        return this.comboInput.nativeElement;
+    }
+
+    /** @hidden @internal */
+    public handleClearItems(event: Event) {
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/simple-combo.component.ts:210]: `handleClear` sounds better IMO

File: projects/igniteui-angular/src/lib/simple-combo/simple-combo.component.ts [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+                this._value = displayText !== args.displayText
+                    ? args.displayText
+                    : this.createDisplayText([args.newSelection], [args.oldSelection]);
+            }
+            this._onChangeCallback(args.newSelection);
+            this._updateInput = true;
+        }
+    }
+
+    private deselectAllItems(ignoreFilter?: boolean): void {
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/simple-combo.component.ts:260]: `deselect` sounds better in context of single selection

**Review by wnvko — CHANGES_REQUESTED** [review]
As a general note check all the items moved from combo to combo.common.

File: projects/igniteui-angular/src/lib/combo/combo-dropdown.component.ts [hunk 13]
```diff
@@ -192,9 +196,13 @@ export class IgxComboDropDownComponent extends IgxDropDownComponent implements I
     private handleEnter() {
         if (this.isAddItemFocused()) {
```
[the remark this answers is not in the document]
Comment (jackofdiamond5) [reply projects/igniteui-angular/src/lib/combo/combo-dropdown.component.ts:197]: Yes, this is the expected behavior.

File: projects/igniteui-angular/src/lib/combo/combo.common.ts (hunk 1 above)
Comment (jackofdiamond5) [reply projects/igniteui-angular/src/lib/combo/combo.common.ts:1271]: Actually it's used in `createDisplayText` which is protected and is being used in the simple combo.

**Review by wnvko — APPROVED** [review]
should add tests for the new component

**Review by Lipata — CHANGES_REQUESTED** [review]
Please add changlog bullet, add component readme and update the main readme with the new component

File: projects/igniteui-angular/migrations/update-13_0_0/changes/members.json [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            "member": "deselectItems",
+            "replaceWith": "deselect",
+            "definedIn": [
+                "IgxComboComponent"
+            ]
+        },
+        {
+            "member": "selectedItems()",
+            "replaceWith": "selected",
```
Comment (Lipata) [inline projects/igniteui-angular/migrations/update-13_0_0/changes/members.json:20]: @wnvko, @jackofdiamond5 isn't it more suitable to be `selection`?

**Review by Lipata — CHANGES_REQUESTED** [review]

File: projects/igniteui-angular/src/lib/simple-combo/README.md [hunk 15]
```diff
@@ -0,0 +1,317 @@
+# igx-simple-combo
+The `igx-simple-combo` is a modification of the `igx-combo` component that allows single selection and has the appropriate UI and behavior for that. It inherits most of the `igx-combo`'s API.
+It provides an editable input used for filtering data and while making use of the IgniteUI for Angular `igx-drop-down` component which is used to display the items in the data set.
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/README.md:3]: `... and while making ...` - I do not get this

File: projects/igniteui-angular/src/lib/simple-combo/README.md [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+Alongside easy filtering and selection of a single item, the control provides grouping and adding of custom values to the dropdown list.
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/README.md:4]: new values are added to the data source not to the dropdown list.

File: projects/igniteui-angular/src/lib/simple-combo/README.md [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+}
+
+public ngAfterViewInit(): void {
+    this.remoteService.getData(this.combo.virtualizationState, (data) => {
+        this.combo.totalItemCount = data.length;
+    });
+}
+
+public dataLoading(evt): void {
+        if (this.prevRequest) {
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/README.md:38]: fix indentation here

File: projects/igniteui-angular/src/lib/simple-combo/README.md [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+<igx-simple-combo [disabled]="true"></igx-simple-combo>
+```
+
+<div class="divider--half"></div>
+
+### Grouping
+Defining a combo's groupKey option will group the items, according to that key.
+
+```html
+<igx-simple-combo [groupKey]="'primaryKey'"></igx-simple-combo>
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/README.md:149]: using `primaryKey` as `groupKey` here could be missleading. Change it with `groupKey` or something else, but not `primaryKey`.

File: projects/igniteui-angular/src/lib/simple-combo/README.md [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+    <ng-template igxComboClearIcon>
+        <igx-icon>clear</igx-icon>
+    </ng-template>
+</igx-simple-combo>
+```
+
+## Keyboard Navigation
+
+When the combo is closed and focused:
+- `ArrowDown` or `Alt` + `ArrowDown` will open the dropdown and will move focus to the selected item, if no selected item is present, the first item in the list wil lbe focused.
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/README.md:246]:
```suggestion
- `ArrowDown` or `Alt` + `ArrowDown` will open the dropdown and will move focus to the selected item, if no selected item is present, the first item in the list will be focused.
```

File: projects/igniteui-angular/src/lib/simple-combo/README.md [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+Setting `[displayDensity]` affects the control's items' and inputs' css properties, most notably heights, padding, font-size.
+
+### Properties
+| Name                     | Description                                       | Type                        |
+|--------------------------|---------------------------------------------------|-----------------------------|
+| `id`                    | combo id                                          | `string`                      |
+| `data`                  | combo data source                                 | `any`                         |
+| `value`                 | combo value                                       | `string`                      |
+| `allowCustomValue`      | enables/disables combo custom value                | `boolean`                     |
+| `valueKey`              | combo value data source property                  | `string`                      |
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/README.md:278]: use the same description as in the component
_Determines which column in the data source is used to determine the value._

File: projects/igniteui-angular/src/lib/simple-combo/README.md [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+| `displayKey`            | combo display data source property                | `string`                      |
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/simple-combo/README.md:279]: use same description as in the component
_Determines which column in the data source is used to determine the display value._

**Review by Lipata — APPROVED** [review]

**Comment by wnvko** [comment]
Issues found in testing:
1. In simple combo with many items (should have scroll in the dropdown) press arrow down + alt, then end then enter. This should select the last item and close the drop-down. Then press ctrl + backspace. This should delete the content on the input and should open the drop down. Now option one - press arrow up -> dropdown closes; option two - press arrow down -> no visible focused item.
2. Open the combo with mouse click. Select any item with mouse click. Click on the clear icon. Dropdown does not open. This is not consistent. If you delete the input content with keyboard, e.g. ctrl + backspace, drop down opens, but clicking on clear button does not open it.
3. In simple combo with many items (should have scroll in the drop-down) open the dropdown with mouse click. Scroll to the last item and select it. Click on the clear button. Open the drop down. The focused item is not visible.
4. The only way to set `groupSortingDirection` in markup is like this `[groupSortingDirection]="2"`. Not sure this is related to this PR.
5. Handle `selectionChanging` event and open empty simple combo. `selectionChanging` is fired with `newSelection === oldSelection === displayText === undefined`.

**Review by wnvko — DISMISSED** [review]

**Review by wnvko — APPROVED** [review]

**Review by Lipata — APPROVED** [review]

**Pull request merged**